source/hp/dof_handler.cc

```diff
@@ -1745,55 +1724,169 @@ namespace hp
     // Resize active_fe_indices vectors. use zero indicator to extend
     for (unsigned int i = 0; i < levels.size(); ++i)
       levels[i]->active_fe_indices.resize(tria->n_raw_cells(i), 0);
+  }
+
 
-    // if a finite element collection has already been set, then
-    // actually try to set active_fe_indices for child cells of
-    // refined cells to the active_fe_index of the mother cell. if no
-    // finite element collection has been assigned yet, then all
-    // indicators are zero anyway, and there is no point trying to set
-    // anything (besides, we would trip over an assertion in
-    // set_active_fe_index)
+  template <int dim, int spacedim>
+  void
+  DoFHandler<dim, spacedim>::pre_refinement_fe_index_update()
+  {
     if (fe_collection.size() > 0)
       {
-        cell_iterator cell = begin(), endc = end();
-        for (; cell != endc; ++cell)
+        Assert(refined_cells_fe_index.empty(), ExcInternalError());
+        Assert(coarsened_cells_fe_index.empty(), ExcInternalError());
+
+        // If the underlying shared::Tria allows artificial cells,
+        // then save the current set of subdomain ids, and set
+        // subdomain ids to the "true" owner of each cell. We later
+        // restore these flags.
+        const parallel::shared::Triangulation<dim, spacedim> *shared_tria =
+          (dynamic_cast<const parallel::shared::Triangulation<dim, spacedim> *>(
+            &(*tria)));
+
+        std::vector<types::subdomain_id> saved_subdomain_ids;
+        if (shared_tria != nullptr)
+          if (shared_tria->with_artificial_cells())
+            {
+              saved_subdomain_ids.resize(shared_tria->n_active_cells());
+
+              typename parallel::shared::Triangulation<dim, spacedim>::
+                active_cell_iterator cell = shared_tria->begin_active(),
+                                     endc = shared_tria->end();
+
+              const std::vector<types::subdomain_id> &true_subdomain_ids =
+                shared_tria->get_true_subdomain_ids_of_cells();
+
+              for (unsigned int index = 0; cell != endc; ++cell, ++index)
+                {
+                  saved_subdomain_ids[index] = cell->subdomain_id();
```

Comment:
here and in the reverse operation, you can write this slightly more elegantly if you do
```
  for (auto &cell : ...)
    saved_subdomain_ids[cell->active_cell_index()] = cell->subdomain_id();
```
This avoids having to explicitly declare the iterators above.